.travis.yml

```diff
@@ -68,8 +14,77 @@ branches:
 cache:
   pip: true
 
-after_success:
-  # only submit code coverage for the newest python version
-  - if [[ $TRAVIS_PYTHON_VERSION == "3.6" ]]; then
-      coveralls;
-    fi
+jobs:
+  include:
+    - stage: test
+      python: 3.5
+      install:
+        - pip install tox
+        - pip install tox-travis
+      script:
+        - tox
+    - &supported_python_build
+      stage: test
+      python: 3.6
+      install:
+        - pip install tox
+        - pip install tox-travis
+        - pip install coveralls
+      script:
+        - tox
+      after_success: coveralls
+    - &docker_build
+      stage: build
+      python: 3.6
+      services:
+        - docker
+      env:
+        - DOCKER_IMAGE=stodevx/cs251-toolkit
+      script:
+        - script/docker-build $DOCKER_IMAGE:$TRAVIS_COMMIT
+    - <<: *supported_python_build
+      stage: deploy
+      after_success: echo "Ready to deploy to PyPI..."
+      deploy:
+        provider: pypi
+        user: "hawkrives"
+        password:
+          secure: kxjTndHBVztrPi1k74zsL2cjbFrBXe5af0HDn/t4w6CUj0yCf0VPwzR0KnOU6CrvRpMxid7MINsgdh3MsOsgnLiLiQ4TSRqEfB9lsmCopDXArNiC5jq0QmZ1P2KHKguJpkWx8TnwjiJycm1UFJU3l4YRfj5Zo4LvpI/64EUoNhICbf2gAQav8Sb2a9/Ov5hpySBYhmwr/tkODOfZS5+00SAs/OP7bU0xQFhUUwnEY4uAiaTIo+tIOIXx/nwj7XundCEwFXiFDYSEKbk+maHWgUIqbZivL94RigVwTds0KJmpurU1Nc670ggGwT96WcCtNZoMTxB47fXoNqzhTAgp2O2C4lsv1xNIrwSdHzyP20wsDQn3zdMz4aJEfSZMwhvN/GJyQFTKbeD++ysRsfUuI5uQho7BOsGMLDfnDOpncNNXI2zKnea9OStsj/u/aJGNmWFcGa70/CHBeSI9Zmv6Og3KY6fGM28bqQUvO6iFDho98xOTHk+vTIcGyut5DDMuKvQh3ISPYZkz7lSqHqDayXr/RxombkpEPszqlUkn2OO38PLujYJzL8nLUN1g/2m2M4y9h6pXV27sU6hDo0E9WTtRhAA04JqpsJ2S8smw2C+yo446klaMNxUkYbxm4j9PI7pYpLWsqlOMbnCz79PSGuzB17ezk6nr8T22TR0Ci2w=
+        on:
+          tags: true
+    - <<: *docker_build
+      stage: deploy
+      deploy:
+        provider: script
+        script: script/docker-deploy $DOCKER_IMAGE:$TRAVIS_COMMIT $DOCKER_IMAGE:HEAD
+        on:
+          branch: master
+    - <<: *docker_build
+      stage: deploy
+      deploy:
+        provider: script
+        script: script/docker-deploy $DOCKER_IMAGE:$TRAVIS_COMMIT $DOCKER_IMAGE:$TRAVIS_TAG
+        on:
+          tags: true
+    - <<: *docker_build
+      stage: deploy
+      deploy:
+        provider: script
+        script: script/docker-deploy $DOCKER_IMAGE:$TRAVIS_COMMIT $DOCKER_IMAGE:$TRAVIS_TAG
+        on:
+          tags: true
+    - <<: *docker_build
+      stage: deploy
+      deploy:
+        provider: script
+        script: script/docker-deploy $DOCKER_IMAGE:$TRAVIS_COMMIT $DOCKER_IMAGE:latest
+        on:
+          tags: true
+          condition: $(script/github-latest-release) == $TRAVIS_TAG
+    - <<: *docker_build
+      stage: deploy
+      deploy:
+        provider: script
+        script: script/docker-deploy $DOCKER_IMAGE:$TRAVIS_COMMIT $DOCKER_IMAGE:$TRAVIS_COMMIT
+        on:
+          branch: travis-*
```

Comment:
Yeah, we used to use the `on` key and a deploy grid in our previous deploys to separate out these different keys.  My goal here was to explore deploying the same tag multiple times in parallel, but I'm realizing now that the increased number of VM's results in greater overhead overall with having to spin up Docker. (which takes ~30s, apparently)